README.md
# Multi Circular Slider

![License: MIT](https://img.shields.io/github/license/mayurnile/multi_circular_slider)

[![Pub Version](https://img.shields.io/pub/v/multi_circular_slider?color=blueviolet)](https://pub.dev/packages/multi_circular_slider)

A highly customizable progress bar for Flutter which helps showing multiple values in single progress bar with different 
colors and also animate it for you.

### Preview

![](./assets/preview.gif)

## Getting Started

- [Installation](#installation)
- [Basic Usage](#basic-usage)
- [MultiCircularSlider parameters](#multicircularslider-parameters)

### Installation

Add

```
dependencies:
  multi_circular_slider: ^1.1.0
```

to your `pubspec.yaml`, and run

```bash
flutter packages get
```

in your project's root directory.

### Basic Usage


Import it to your project file

```dart
import 'package:multi_circular_slider/multi_circular_slider.dart';
```

And add it in its most basic form like it:

```dart
MultiCircularSlider(
    size: MediaQuery.of(context).size.width * 0.8,
    progressBarType: MultiCircularSliderType.circular,
    values: [0.2, 0.1, 0.3, 0.25],
    colors: [Color(0xFFFD1960), Color(0xFF29D3E8), Color(0xFF18C737), Color(0xFFFFCC05)],
    showTotalPercentage: true,
),

```

There are additional optional parameters one can initialize the slider with.

```dart
MultiCircularSlider(
    size: MediaQuery.of(context).size.width * 0.8,
    progressBarType: MultiCircularSliderType.circular, // the type of indictor you want circular or linear
    values: [0.2, 0.1, 0.3, 0.25],
    colors: [Color(0xFFFD1960), Color(0xFF29D3E8), Color(0xFF18C737), Color(0xFFFFCC05)],
    showTotalPercentage: true, // to display total percentage in center
    label: 'This is label text', // label to display below percentage
    animationDuration: const Duration(milliseconds: 500), // duration of animation
    animationCurve: Curves.easeIn, // smoothness of animation
    innerIcon: Icon(Icons.integration_instructions), // to display some icon related to text
    innerWidget: Text('96%'), // to show custom innerWidget (to display set showTotalPercentage to false)
    trackColor: Colors.white, // to change color of track
    progressBarWidth: 52.0, // to change width of progressBar
    trackWidth: 52.0, // to change width of track
    labelTextStyle: TextStyle(), // to change TextStyle of label
    percentageTextStyle: TextStyle(), // to change TextStyle of percentage
);
```

### MultiCircularSlider parameters

| Parameter | DataType | Default Value | Description |
| ------ | ------ | ------ | ------ |
| size | Size | | the space widget should take up on screen |
| values | List<double> | | pass different percentages you want to show which sum up to 1.0 or less |
| colors | List<Color> | | different colors which you want to give to the progress bars (NOTE: length of `values` & `colors` should be same) |
| progressBarType | MultiCircularSliderType | | to specify the type of progress indicator you want |
| showTotalPercentage | bool | true | whether to show total percentage in center or not |
| label | String | | any label text which you want to show below total percentage |
| animationDuration | Duration | Duration(milliseconds: 500) | the duration you want for the animation |
| animationCurve | Curve | Curves.easeInOutCubic | the curve you want for animation |
| innerIcon | Widget | | the icon which you can display above the total percentage text |
| innerWidget | Widget | | the widget you want to show inside the circular progress bar (NOTE : innerWidget will only de displayed if showTotalPercentage is false) |
| innerWidgetPadding | EdgeInsets | size * 0.15 | padding between inner widget and progressBar |
| trackColor | Color | Colors.grey | color of the track of progressBar |
| progressBarWidth | double | trackWidth | stroke width of the progressBar |
| trackWidth | double | 32.0 | stroke width of the progressBar track |
| labelTextStyle | TextStyle | | TextStyle which you want to give to label |
| percentageTextStyle | TextStyle | | TextStyle which you want to give to percentage |
| startAngle | double | 180.0 | starting angle of circular progress bar |
| sweepAngle | double | 180.0 | sweep angle of circular progress bar |


CHANGELOG.md
## [0.0.1] - 7th December 2020.

* First release of Multi Circular Slider for Flutter.

## [0.0.2] - 7th December 2020.

* Updated Readme.

## [0.0.3] - 7th December 2020.

* Re-formatted some code for readability.

## [1.0.0] - 16th May 2022.
* Migrated to Null Safety
* Solved issue track width not getting adjusted

## [1.1.0] - 10th Aug 2022.
* Added support for Linear Progress Indicator

## [1.1.1] - 28th Apr 2023.
* Fixed bug (progressBarWidth not working).
* Added support for specifying starting angle and sweep angle of Circular Progress Bar.
* Added support for specifying innerWidget padding.
* Updated Readme.


example/multi_progress_bar_demo/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:multi_progress_bar_demo/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/circular_slider.dart
library circular_slider;

import 'dart:math' as math;

import 'package:flutter/material.dart';

part 'circular_progress_bar_painter.dart';
part 'linear_progress_bar_painter.dart';

enum MultiCircularSliderType {
  // to create a linear progress bar
  linear,
  // to create a circular progress bar
  circular,
}

class MultiCircularSlider extends StatefulWidget {
  final double size;
  final MultiCircularSliderType progressBarType;
  final double trackWidth;
  final double? progressBarWidth;
  final List<double> values;
  final List<Color> colors;
  final Color trackColor;
  final Duration animationDuration;
  final Curve animationCurve;
  final Widget? innerWidget;
  final Widget? innerIcon;
  final bool showTotalPercentage;
  final String? label;
  final TextStyle? percentageTextStyle;
  final TextStyle? labelTextStyle;
  final double startAngle;
  final double sweepAngle;
  final EdgeInsets? innerWidgetPadding;

  ///A widget to make multiple progress bar to show
  ///multiple values in a single bar
  ///
  ///For example to show storage space taken up by different types of files in File Manager
  ///
  ///
  ///`values` pass different percentages you want to show which sum up to 1.0 or less
  ///
  ///
  ///
  ///`colors` different colors which you want to give to the progress bars
  ///
  ///NOTE : Length of values should be equal to length of colors
  ///
  ///
  ///
  ///`size` the space widget should take up on screen
  ///
  ///
  ///
  ///`startAngle` the start angle of the circular progress bar
  ///
  ///defualt value is set to 180.0
  ///
  ///
  ///`sweepAngle` the sweep angle of the circular progress bar
  ///
  ///defualt value is set to 180.0
  ///
  ///
  ///`progressBarType` the type of progress bar you want to show
  ///
  ///default value is `trackWidth`
  ///
  ///`trackWidth` stroke width of the progressBar track
  ///
  ///default values is set to 32.0
  ///
  ///
  ///
  ///`progressBarWidth` stroke width of the progressBar
  ///
  ///default values is set to 32.0
  ///
  ///
  ///`trackColor` color of the track of progressBar
  ///
  ///default values is set to Colors.grey
  ///
  ///
  ///`animationDuration` the duration you want for the animation
  ///
  ///default values is set to 1000 milliseconds
  ///
  ///
  ///
  ///`animationCurve` the curve you want for animation
  ///
  ///default values is set to Curves.easeInOutCubic
  ///
  ///
  ///
  ///`innerWidget` the widget you want to show inside the circular progress bar
  ///
  ///
  ///
  ///`innerIcon` the icon which you can display above the total percentage text
  ///
  ///
  ///
  ///`innerWidgetPadding` padding between `innerWidget` and circular progress bar
  ///
  ///default value is 15% of `size` on all sides
  ///
  ///
  ///
  ///`showTotalPercentage` whether to show total percentage in center or not
  ///
  ///default values is set to true
  ///
  ///
  ///
  ///`label` any label text which you want to show below total percentage
  ///
  ///
  ///
  ///`labelTextStyle` TextStyle which you want to give to label
  ///
  ///
  ///
  ///`percentageTextStyle` TextStyle which you want to give to percentage
  ///
  const MultiCircularSlider({
    Key? key,
    required this.values,
    required this.colors,
    required this.size,
    required this.progressBarType,
    this.trackWidth = 32.0,
    this.progressBarWidth,
    this.trackColor = Colors.grey,
    this.animationDuration = const Duration(milliseconds: 1000),
    this.animationCurve = Curves.easeInOutCubic,
    this.innerWidget,
    this.innerIcon,
    this.showTotalPercentage = true,
    this.label,
    this.labelTextStyle,
    this.percentageTextStyle,
    this.startAngle = 180.0,
    this.sweepAngle = 180.0,
    this.innerWidgetPadding,
  }) : super(key: key);

  @override
  MultiCircularSliderState createState() => MultiCircularSliderState();
}

class MultiCircularSliderState extends State<MultiCircularSlider>
    with SingleTickerProviderStateMixin {
  //for animation
  late AnimationController _controller;
  late Animation<double> _animation;

  //for calculation
  double percentage = 0.0;

  @override
  void initState() {
    super.initState();

    //calculating total percentage
    for (final value in widget.values) {
      percentage += value;
    }

    //initializing controller
    _controller =
        AnimationController(vsync: this, duration: widget.animationDuration);

    //initializing animation
    _animation = Tween(
      begin: 0.0,
      end: 1.0,
    ).animate(
        CurvedAnimation(parent: _controller, curve: widget.animationCurve));

    //starting animation
    _controller.forward();
  }

  @override
  void dispose() {
    //dispose controller to avoid memory leaks
    _controller.dispose();

    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    //actual widget starts here
    return AnimatedBuilder(
      animation: _animation,
      builder: (BuildContext context, Widget? child) {
        //custom painter to draw multiple progress bar
        return widget.progressBarType == MultiCircularSliderType.circular
            ? _buildCircularProgressBar()
            : _buildLinearProgressBar();
      },
      child: widget.innerWidget ?? const SizedBox.shrink(),
    );
  }

  /// Builder Functions
  ///
  ///
  Widget _buildCircularProgressBar() => CustomPaint(
        painter: _buildProgressBarPainter(),
        //inner widget with shadow
        child: Container(
          height: widget.size,
          width: widget.size,
          //Affects inner widget
          padding:
              widget.innerWidgetPadding ?? EdgeInsets.all(widget.size * 0.15),
          child: widget.innerWidget ??
              Container(
                // padding: const EdgeInsets.all(32.0),
                padding: const EdgeInsets.all(4.0),
                decoration: BoxDecoration(
                  color: Colors.white,
                  shape: BoxShape.circle,
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(0.07),
                      blurRadius: 10.0,
                      offset: const Offset(0.0, -18.0),
                    ),
                  ],
                ),
                child: widget.showTotalPercentage
                    ? Column(
                        children: [
                          //inner icon
                          _buildInnerIcon(),
                          //total percentage
                          _buildTotalPercentage(),
                          //spacing
                          const SizedBox(height: 8.0),
                          //text
                          _buildTextWidget(),
                        ],
                      )
                    : const SizedBox.shrink(),
              ),
        ),
      );

  Widget _buildLinearProgressBar() => Column(
        mainAxisSize: MainAxisSize.min,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          // inner widget
          widget.innerWidget ?? const SizedBox.shrink(),
          // total percentage
          widget.showTotalPercentage
              ? Column(
                  children: [
                    //inner icon
                    _buildInnerIcon(),
                    //total percentage
                    _buildTotalPercentage(),
                    //spacing
                    const SizedBox(height: 8.0),
                    //text
                    _buildTextWidget(),
                  ],
                )
              : const SizedBox.shrink(),
          // linear progress bar
          CustomPaint(
            painter: _buildProgressBarPainter(),
            size: Size(widget.size, widget.size),
          ),
        ],
      );

  CustomPainter _buildProgressBarPainter() =>
      widget.progressBarType == MultiCircularSliderType.circular
          ? CircularProgressBarPainter(
              size: widget.size,
              startAngle: widget.startAngle,
              baseAngle: widget.sweepAngle,
              values: List.generate(widget.values.length,
                  (index) => widget.values[index] * _animation.value),
              colors: widget.colors,
              progressBarWidth: widget.progressBarWidth ?? widget.trackWidth,
              trackColor: widget.trackColor,
              trackWidth: widget.trackWidth,
            )
          : LinearProgressBarPainter(
              size: widget.size,
              values: List.generate(widget.values.length,
                  (index) => widget.values[index] * _animation.value),
              colors: widget.colors,
              // progressBarWidth: widget.progressBarWidth,
              progressBarWidth: widget.progressBarWidth ?? widget.trackWidth,
              trackColor: widget.trackColor,
              trackWidth: widget.trackWidth,
            );

  Widget _buildInnerIcon() => widget.innerIcon ?? const SizedBox.shrink();

  Widget _buildTotalPercentage() => Text(
        '${(percentage * _animation.value * 100).ceil()}%',
        textAlign: TextAlign.center,
        style: widget.percentageTextStyle ??
            const TextStyle(
              fontWeight: FontWeight.w900,
              fontSize: 28.0,
              color: Color(0xFF012C61),
            ),
      );

  Widget _buildTextWidget() => widget.label != null
      ? FittedBox(
          child: Text(
            widget.label!,
            style: widget.labelTextStyle ??
                const TextStyle(
                  fontWeight: FontWeight.w600,
                  fontSize: 22.0,
                  color: Color(0xFF939AA4),
                ),
          ),
        )
      : const SizedBox.shrink();
}


lib/src/progress_bar_painter.dart
part of circular_slider;

/// CustomPainter class to draw progress bar curves
class ProgressBarPainter extends CustomPainter {
  //size of the widget
  final double size;

  //width of the track
  final double trackWidth;

  //width of the progressbar
  final double progressBarWidth;

  //list of total values
  final List<double> values;

  //list of colors
  final List<Color> colors;

  //color of track
  final Color trackColor;

  //constructor
  ProgressBarPainter({
    required this.size,
    this.trackWidth = 32.0,
    this.progressBarWidth = 32.0,
    required this.values,
    required this.colors,
    this.trackColor = Colors.grey,
  });

  @override
  void paint(Canvas canvas, Size size) {
    //base angle of curve
    const double baseAngle = 180.0;
    //actual starting angle of arc
    final double startAngle = degreeToRadians(baseAngle);
    //total length of the track of progressbar
    final double trackSweepAngle = degreeToRadians(baseAngle);

    //shadow paint
    final shadowPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = Colors.black.withOpacity(0.1)
      ..maskFilter = const MaskFilter.blur(
        BlurStyle.normal,
        10.0,
      )
      ..strokeWidth = trackWidth;
    //draw shadow
    drawCurve(
      canvas,
      size,
      startAngle,
      trackSweepAngle,
      shadowPaint,
    );

    //progress bar track paint
    final trackPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = trackColor
      ..strokeWidth = trackWidth;
    //progress bar track curve
    drawCurve(
      canvas,
      size,
      startAngle,
      trackSweepAngle,
      trackPaint,
    );

    //length of list
    int length = values.length;
    double totalPercentage = 0.0;

    //to store values in reverse order
    List<Shader> progressBars = [];

    //iterating through list for calculating values
    for (int i = 0; i < length; i++) {
      double percentage = baseAngle * values[i];
      totalPercentage += percentage;
      double sweepAngle = degreeToRadians(totalPercentage);

      //progress bar paint
      final progressBarPaint = Paint()
        // ..shader = progressBarGradient.createShader(progressBarRect)
        ..strokeCap = StrokeCap.round
        ..style = PaintingStyle.stroke
        ..color = colors[i]
        ..strokeWidth = trackWidth;

      //adding values to list
      progressBars.insert(
        0,
        Shader(
          startAngle: startAngle,
          sweepAngle: sweepAngle,
          paint: progressBarPaint,
        ),
      );
    }

    //drawing actual progress bars
    for (final progressBar in progressBars) {
      drawCurve(
        canvas,
        size,
        progressBar.startAngle,
        progressBar.sweepAngle,
        progressBar.paint,
      );
    }
  }

  ///[Logic for drawing curve]
  drawCurve(
    Canvas canvas,
    Size size,
    double startAngle,
    double sweepAngle,
    Paint paint,
  ) {
    double radius = size.width / 2;
    Offset center = Offset(
      size.width / 2,
      size.height / 2,
    );

    canvas.drawArc(
      Rect.fromCircle(
        center: center,
        radius: radius,
      ),
      startAngle,
      sweepAngle,
      false,
      paint,
    );
  }

  ///[Logic for converting degree to radians]
  degreeToRadians(double degree) {
    return degree * (math.pi / 180);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

/// class for storing values of the ProgressBar
class Shader {
  final double startAngle;
  final double sweepAngle;
  final Paint paint;

  Shader({
    required this.startAngle,
    required this.sweepAngle,
    required this.paint,
  });
}


lib/src/linear_progress_bar_painter.dart
part of circular_slider;

/// CustomPainter class to draw linear progress bar
class LinearProgressBarPainter extends CustomPainter {
  //size of the widget
  final double size;

  //width of the track
  final double trackWidth;

  //width of the progressbar
  final double progressBarWidth;

  //list of total values
  final List<double> values;

  //list of colors
  final List<Color> colors;

  //color of track
  final Color trackColor;

  //constructor
  LinearProgressBarPainter({
    required this.size,
    this.trackWidth = 32.0,
    required this.progressBarWidth,
    required this.values,
    required this.colors,
    this.trackColor = Colors.grey,
  });

  @override
  void paint(Canvas canvas, Size size) {
    //shadow paint
    final shadowPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = Colors.black.withOpacity(0.1)
      ..maskFilter = const MaskFilter.blur(
        BlurStyle.normal,
        10.0,
      )
      ..strokeWidth = trackWidth;
    // ..strokeWidth = math.max(trackWidth, progressBarWidth);

    //draw shadow
    canvas.drawLine(
      const Offset(0, 0),
      Offset(size.width, 0),
      shadowPaint,
    );

    //progress bar track paint
    final trackPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = trackColor
      ..strokeWidth = trackWidth;

    //draw progrwss bar track
    canvas.drawLine(
      const Offset(0, 0),
      Offset(size.width, 0),
      trackPaint,
    );

    //length of list
    int length = values.length;
    double totalPercentage = 0.0;

    //to store values in reverse order
    List<LinearShader> progressBarsPainters = [];

    //iterating through list for calculating values
    for (int i = 0; i < length; i++) {
      totalPercentage += values[i];

      //progress bar paint
      final progressBarPaint = Paint()
        ..strokeCap = StrokeCap.round
        ..style = PaintingStyle.stroke
        ..color = colors[i]
        ..strokeWidth = progressBarWidth;

      //adding values to list
      progressBarsPainters.insert(
        0,
        LinearShader(
          width: size.width * totalPercentage,
          paint: progressBarPaint,
        ),
      );
    }

    //drawing actual progress bars
    for (final progressBarPaint in progressBarsPainters) {
      canvas.drawLine(
        const Offset(0, 0),
        Offset(progressBarPaint.width, 0),
        progressBarPaint.paint,
      );
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

/// class for storing values of the ProgressBar
class LinearShader {
  final double width;
  final Paint paint;

  LinearShader({
    required this.width,
    required this.paint,
  });
}


lib/src/circular_progress_bar_painter.dart
part of circular_slider;

/// CustomPainter class to draw progress bar curves
class CircularProgressBarPainter extends CustomPainter {
  //size of the widget
  final double size;

  //width of the track
  final double trackWidth;

  //width of the progressbar
  final double progressBarWidth;

  //list of total values
  final List<double> values;

  //list of colors
  final List<Color> colors;

  //color of track
  final Color trackColor;

  //color of track
  final double baseAngle;

  //color of track
  final double startAngle;

  //constructor
  CircularProgressBarPainter({
    required this.size,
    this.trackWidth = 32.0,
    required this.progressBarWidth,
    required this.values,
    required this.colors,
    this.trackColor = Colors.grey,
    required this.baseAngle,
    required this.startAngle,
  });

  @override
  void paint(Canvas canvas, Size size) {
    //actual starting angle of arc
    final double startAngleInRadians = degreeToRadians(startAngle);
    //total length of the track of progressbar
    final double trackSweepAngle = degreeToRadians(baseAngle);

    //shadow paint
    final shadowPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = Colors.black.withOpacity(0.1)
      ..maskFilter = const MaskFilter.blur(
        BlurStyle.normal,
        10.0,
      )
      ..strokeWidth = trackWidth;
    // ..strokeWidth = math.max(trackWidth, progressBarWidth);
    //draw shadow
    drawCurve(
      canvas,
      size,
      startAngleInRadians,
      trackSweepAngle,
      shadowPaint,
    );

    //progress bar track paint
    final trackPaint = Paint()
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke
      ..color = trackColor
      ..strokeWidth = trackWidth;
    //progress bar track curve
    drawCurve(
      canvas,
      size,
      startAngleInRadians,
      trackSweepAngle,
      trackPaint,
    );

    //length of list
    int length = values.length;
    double totalPercentage = 0.0;

    //to store values in reverse order
    List<CircularShader> progressBars = [];

    //iterating through list for calculating values
    for (int i = 0; i < length; i++) {
      double percentage = baseAngle * values[i];
      totalPercentage += percentage;
      double sweepAngle = degreeToRadians(totalPercentage);

      //progress bar paint
      final progressBarPaint = Paint()
        ..strokeCap = StrokeCap.round
        ..style = PaintingStyle.stroke
        ..color = colors[i]
        ..strokeWidth = progressBarWidth;

      //adding values to list
      progressBars.insert(
        0,
        CircularShader(
          startAngle: startAngleInRadians,
          sweepAngle: sweepAngle,
          paint: progressBarPaint,
        ),
      );
    }

    //drawing actual progress bars
    for (final progressBar in progressBars) {
      drawCurve(
        canvas,
        size,
        progressBar.startAngle,
        progressBar.sweepAngle,
        progressBar.paint,
      );
    }
  }

  ///[Logic for drawing curve]
  drawCurve(
    Canvas canvas,
    Size size,
    double startAngle,
    double sweepAngle,
    Paint paint,
  ) {
    double radius = size.width / 2;
    Offset center = Offset(
      size.width / 2,
      size.height / 2,
    );

    canvas.drawArc(
      Rect.fromCircle(
        center: center,
        radius: radius,
      ),
      startAngle,
      sweepAngle,
      false,
      paint,
    );
  }

  ///[Logic for converting degree to radians]
  degreeToRadians(double degree) {
    return degree * (math.pi / 180);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}

/// class for storing values of the ProgressBar
class CircularShader {
  final double startAngle;
  final double sweepAngle;
  final Paint paint;

  CircularShader({
    required this.startAngle,
    required this.sweepAngle,
    required this.paint,
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:multi_circular_slider/multi_circular_slider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Multi Progress Bar',
      home: MyTestWidget(),
    );
  }
}

class MyTestWidget extends StatelessWidget {
  const MyTestWidget({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Center(
          child: MultiCircularSlider(
            size: MediaQuery.of(context).size.width * 0.8,
            progressBarType: MultiCircularSliderType.circular,
            values: const [0.2, 0.1, 0.3, 0.28],
            colors: const [
              Color(0xFFFD1960),
              Color(0xFF29D3E8),
              Color(0xFF18C737),
              Color(0xFFFFCC05)
            ],
            showTotalPercentage: true,
            label: 'This is label text',
            animationDuration: const Duration(milliseconds: 1000),
            animationCurve: Curves.easeInOutCirc,
            innerIcon: const Icon(Icons.integration_instructions),
            trackColor: Colors.white,
            progressBarWidth: 38,
            trackWidth: 38,
            // startAngle: 180,
            // sweepAngle: 270,
            // innerWidgetPadding: const EdgeInsets.all(40),
            labelTextStyle: const TextStyle(),
            percentageTextStyle: const TextStyle(),
          ),
        ),
      ),
    );
  }
}
